src/hardened_only.rs

```diff
@@ -73,48 +87,61 @@ impl<C: Context> HardenedOnlyKey<C> {
             for input in ikm {
                 I.update(input);
             }
-            I.finalize().as_bytes().try_into().unwrap()
+            I.finalize().as_bytes().try_into().expect("64-byte output")
         };
+        Self::derive_from(&I)
+    }
 
-        let (I_L, I_R) = I.split_at(32);
-
-        // I_L is used as the master secret key sk_m.
-        let sk_m = I_L.try_into().unwrap();
-
-        // I_R is used as the master chain code c_m.
-        let c_m = ChainCode::new(I_R.try_into().unwrap());
+    /// Derives a child key from a parent key at a given index and empty tag.
+    ///
+    /// This is a convenience function equivalent to
+    /// `self.derive_child_with_tag(index, &[])`.
+    pub fn derive_child(&self, index: ChildIndex) -> Self {
+        self.derive_child_with_tag(index, &[])
+    }
 
-        Self {
-            sk: sk_m,
-            chain_code: c_m,
-            _context: PhantomData,
-        }
+    /// Derives a child key from a parent key at a given index and (possibly empty)
+    /// tag.
+    ///
+    /// Defined in [ZIP 32: Hardened-only child key derivation][ckdh].
+    ///
+    /// [ckdh]: https://zips.z.cash/zip-0032#hardened-only-child-key-derivation
+    pub fn derive_child_with_tag(&self, index: ChildIndex, tag: &[u8]) -> Self {
+        Self::derive_from(&self.ckdh_internal(index, 0, tag))
     }
 
-    /// Derives a child key from a parent key at a given index.
+    /// Defined in [ZIP 32: Hardened-only child key derivation][ckdh].
     ///
-    /// Defined in [ZIP32: Hardened-only child key derivation][ckdh].
+    /// This returns `I` rather than `(I_L, I_R)` so that we don't have
+    /// to re-concatenate the pieces, e.g. when using it in
+    /// [`crate::registered::SecretKey::derive_child_cryptovalue`].
     ///
     /// [ckdh]: https://zips.z.cash/zip-0032#hardened-only-child-key-derivation
-    pub fn derive_child(&self, index: ChildIndex) -> Self {
-        // I := PRF^Expand(c_par, [Context.CKDDomain] || sk_par || I2LEOSP(i))
-        let I: [u8; 64] = C::CKD_DOMAIN.with(
+    pub(crate) fn ckdh_internal(&self, index: ChildIndex, lead: u8, tag: &[u8]) -> [u8; 64] {
+        // One of these depending on lead and tag:
+        // - I := PRF^Expand(c_par, [Context.CKDDomain] || sk_par || I2LEOSP(i))
+        // - I := PRF^Expand(c_par, [Context.CKDDomain] || sk_par || I2LEOSP(i) || [lead] || tag)
+        C::CKD_DOMAIN.with(
             self.chain_code.as_bytes(),
             &self.sk,
             &index.index().to_le_bytes(),
-        );
+            &[lead],
+            tag,
+        )
+    }
 
+    fn derive_from(I: &[u8; 64]) -> Self {
```

Comment:
nit, nonblocking: this isn't actually doing any derivation, so I might have named it `from_bytes`.